header/p2p/peer_tracker.go

```diff
@@ -190,3 +194,28 @@ func (p *peerTracker) stop() {
 		<-p.done
 	}
 }
+
+// blockPeer blocks a peer on the networking level and removes it from the local cache.
+func (p *peerTracker) blockPeer(pID peer.ID) {
+	// add peer to the blacklist, so we can't connect to it in the future.
+	err := p.connGater.BlockPeer(pID)
+	if err != nil {
+		log.Errorw("header/p2p: blocking peer failed", "err", err, "pID", pID)
+	}
+	// close connections to peer.
+	err = p.host.Network().ClosePeer(pID)
+	if err != nil {
+		log.Errorw("header/p2p: closing connection with peer failed", "err", err, "pID", pID)
+	}
+
+	log.Debugw("header/p2p: blocked peer", "pID", pID)
+
+	p.peerLk.Lock()
+	defer p.peerLk.Unlock()
+	// remove peer from cache.
+	if _, ok := p.trackedPeers[pID]; ok {
+		delete(p.trackedPeers, pID)
+	} else if _, ok = p.disconnectedPeers[pID]; ok {
+		delete(p.disconnectedPeers, pID)
+	}
```

Comment:
In Go, you don't need to check if the entry is on the map or not, because it does so internally. If there is nothing remove, the delete is just noop